src/engine/devlinks.rs

```diff
@@ -17,104 +17,153 @@ use super::engine::DEV_PATH;
     if let Err(err) = fs::create_dir(DEV_PATH) {
         if err.kind() != ErrorKind::AlreadyExists {
-            return Err(From::from(err));
+            warn!(
+                "Unable to create {} directory, expect more errors {:?}",
+                DEV_PATH, err
+            );
         }
     }
-
-    Ok(())
 }
 
-/// Set up directories and symlinks under /dev/stratis based on current
-/// config. Clear out any directory or file that doesn't correspond to a pool
-/// or filesystem.
+/// Setup the pool directory and the symlinks in /dev/stratis for the specified pool and filesystems
+/// it contains.
 // Don't just remove and recreate everything in case there are processes
 // (e.g. user shells) with the current working directory within the tree.
-pub fn setup_devlinks<'a, I: Iterator<Item = &'a (Name, PoolUuid, &'a Pool)>>(
-    pools: I,
-) -> StratisResult<()> {
-    let mut existing_dirs = fs::read_dir(DEV_PATH)?
-        .map(|dir_e| dir_e.and_then(|d| Ok(d.file_name().into_string().expect("Unix is utf-8"))))
-        .collect::<Result<HashSet<_>, _>>()?;
-
-    for &(ref pool_name, _, pool) in pools {
-        if !existing_dirs.remove(&pool_name.to_owned()) {
-            pool_added(pool_name)?;
-        }
-
+pub fn setup_pool_devlinks(pool_name: &str, pool: &Pool) -> () {
+    match || -> StratisResult<()> {
         let pool_path = pool_directory(pool_name);
 
+        if !pool_path.exists() {
+            pool_added(pool_name);
+        }
+
         let mut existing_files = fs::read_dir(pool_path)?
             .map(|dir_e| {
                 dir_e.and_then(|d| Ok(d.file_name().into_string().expect("Unix is utf-8")))
             })
             .collect::<Result<HashSet<_>, _>>()?;
 
         for (fs_name, _, fs) in pool.filesystems() {
-            filesystem_added(pool_name, &fs_name, &fs.devnode())?;
+            filesystem_added(pool_name, &fs_name, &fs.devnode());
             existing_files.remove(&fs_name.to_owned());
         }
 
         for leftover in existing_files {
-            filesystem_removed(pool_name, &leftover)?;
+            filesystem_removed(pool_name, &leftover);
         }
-    }
 
-    for leftover in existing_dirs {
-        pool_removed(&Name::new(leftover))?
-    }
+        Ok(())
+    }() {
+        Err(e) => {
+            warn!(
+                "setup_pool_devlinks failed for /dev/stratis/{}, reason {:?}",
+                pool_name, e
+            );
+            ()
+        }
+        Ok(_) => (),
+    };
+}
 
-    Ok(())
+/// Set up directories and symlinks under /dev/stratis based on current
+/// config. Clear out any directory or file that doesn't correspond to a pool
+/// or filesystem.
+// Don't just remove and recreate everything in case there are processes
+// (e.g. user shells) with the current working directory within the tree.
+pub fn setup_devlinks<'a, I: Iterator<Item = &'a (Name, PoolUuid, &'a Pool)>>(pools: I) -> () {
+    match || -> StratisResult<()> {
```

Comment:
This is a good way to "catch" errors from multiple sources and handle them in one spot, but I now only see one thing ("?") in the closure that can fail. If we just warn on that directly, we can do away with the closure.

Reply:
Actually there is a couple of `?'s` in the function `setup_devlinks` which makes it just like `setup_pool_devlinks`.